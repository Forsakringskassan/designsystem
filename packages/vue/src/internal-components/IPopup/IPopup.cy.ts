import { type DefineComponent, defineComponent } from "vue";
import { IPopupPageObject } from "../../pageobject";
import IPopup from "./IPopup.vue";

const popup = new IPopupPageObject();
const popupButtonId = "#open-popup";

const VIEWPORT = {
    DESKTOP: { width: 1024, height: 600 },
    MOBILE: { width: 639, height: 600 },
    SHORT: { width: 1024, height: 300 }, // low enough height to result in no valid popup candidate
};

function setViewport(viewPort: { height: number; width: number }): void {
    cy.viewport(viewPort.width, viewPort.height);
}

function createComponent(
    popupTemplate: string = defaultPopupTemplate,
): DefineComponent {
    return defineComponent({
        template: /* HTML */ `
            <div style="min-height: 95vh">
                <button
                    type="button"
                    id="open-popup"
                    ref="anchor"
                    @click="isOpen=!isOpen"
                >
                    Toggle popup
                </button>
                ${popupTemplate}
            </div>
        `,
        components: {
            IPopup,
        },
        data() {
            return {
                isOpen: false,
            };
        },
        methods: {
            myFocusElementCallback() {
                return document.querySelector("#focus-callback-target");
            },
        },
    });
}

const defaultPopupTemplate = /* HTML */ `
    <i-popup :isOpen="isOpen" :anchor="$refs.anchor">
        <div style="height: 300px; width: 250px;">
            <span> POPUP CONTENT </span>
        </div>
    </i-popup>
`;

describe("open popup", () => {
    describe("classes", () => {
        describe("default", () => {
            it("should add `popup--overlay` class as default at desktop size", () => {
                setViewport(VIEWPORT.DESKTOP);
                const component = createComponent();
                cy.mount(component);
                cy.get(popupButtonId).click();
                popup.el().should("have.class", "popup--overlay");
            });

            it("should add `popup--inline` class as default at mobile size", () => {
                setViewport(VIEWPORT.MOBILE);
                const component = createComponent();
                cy.mount(component);
                cy.get(popupButtonId).click();
                popup.el().should("have.class", "popup--inline");
            });

            it("should add `popup--inline` class as default at short height", () => {
                setViewport(VIEWPORT.SHORT);
                const component = createComponent();
                cy.mount(component);
                cy.get(popupButtonId).click();
                popup.el().should("have.class", "popup--inline");
            });
        });

        describe("`inline` prop", () => {
            function setInlineTemplate(inline: string): string {
                return /* HTML */ `
                    <i-popup
                        :isOpen="isOpen"
                        :anchor="$refs.anchor"
                        inline="${inline}"
                    >
                        <div style="height: 300px; width: 250px;">
                            <span> POPUP CONTENT </span>
                        </div>
                    </i-popup>
                `;
            }

            it("should add `popup--inline` class when set to `always`", () => {
                setViewport(VIEWPORT.DESKTOP);
                const template = setInlineTemplate("always");
                const component = createComponent(template);
                cy.mount(component);
                cy.get(popupButtonId).click();
                popup.el().should("have.class", "popup--inline");
            });

            it("should add `popup--overlay` class when set to `never`", () => {
                setViewport(VIEWPORT.DESKTOP);
                const template = setInlineTemplate("never");
                const component = createComponent(template);
                cy.mount(component);
                cy.get(popupButtonId).click();
                popup.el().should("have.class", "popup--overlay");
            });

            it("should add `popup--overlay` class when set to `auto` and at desktop size", () => {
                setViewport(VIEWPORT.DESKTOP);
                const template = setInlineTemplate("auto");
                const component = createComponent(template);
                cy.mount(component);
                cy.get(popupButtonId).click();
                popup.el().should("have.class", "popup--overlay");
            });

            it("should add `popup--inline` class when set to `auto` and at mobile size", () => {
                setViewport(VIEWPORT.MOBILE);
                const template = setInlineTemplate("auto");
                const component = createComponent(template);
                cy.mount(component);
                cy.get(popupButtonId).click();
                popup.el().should("have.class", "popup--inline");
            });

            it("should add `popup--inline` class when set to `auto` and at short height", () => {
                setViewport(VIEWPORT.SHORT);
                const template = setInlineTemplate("auto");
                const component = createComponent(template);
                cy.mount(component);
                cy.get(popupButtonId).click();
                popup.el().should("have.class", "popup--inline");
            });
        });

        describe("always-inline prop", () => {
            it("should add `popup--inline` class when prop `always-inline` is set", () => {
                setViewport(VIEWPORT.DESKTOP);
                const template = /* HTML */ `
                    <i-popup
                        :isOpen="isOpen"
                        :anchor="$refs.anchor"
                        always-inline
                    >
                        <span> POPUP CONTENT </span>
                    </i-popup>
                `;
                const component = createComponent(template);
                cy.mount(component);
                cy.get(popupButtonId).click();
                popup.el().should("have.class", "popup--inline");
            });
        });
    });

    describe("focus", () => {
        beforeEach(() => {
            setViewport(VIEWPORT.DESKTOP);
        });

        it("should focus on first tabbable element by default", () => {
            const template = /* HTML */ `
                <i-popup :isOpen="isOpen" :anchor="$refs.anchor">
                    <span> POPUP CONTENT </span>
                    <!-- The button that will receive focus -->
                    <button type="button" id="b1">Focus me!</button>
                </i-popup>
            `;
            const component = createComponent(template);
            cy.mount(component);
            cy.get(popupButtonId).click();
            cy.get("#b1").should("have.focus");
        });

        it("should focus on specific tabbable element when callback is used", () => {
            const template = /* HTML */ `
                <i-popup
                    :isOpen="isOpen"
                    :anchor="$refs.anchor"
                    :focusElement="myFocusElementCallback"
                >
                    <span> POPUP CONTENT </span>
                    <button type="button" id="b1">Button 1</button>
                    <!-- The button that will receive focus -->
                    <button type="button" id="focus-callback-target">
                        Button 2
                    </button>
                </i-popup>
            `;
            const component = createComponent(template);
            cy.mount(component);
            cy.get(popupButtonId).click();
            cy.get("#focus-callback-target").should("have.focus");
        });

        it("should skip setting focus if `setFocus` is false", () => {
            const template = /* HTML */ `
                <i-popup
                    :isOpen="isOpen"
                    :anchor="$refs.anchor"
                    :setFocus="false"
                >
                    <span> POPUP CONTENT </span>
                    <button type="button" id="b1">Button 1</button>
                </i-popup>
            `;
            const component = createComponent(template);
            cy.mount(component);
            cy.get(popupButtonId).click();
            cy.get(popupButtonId).should("have.focus");
        });
    });
});
