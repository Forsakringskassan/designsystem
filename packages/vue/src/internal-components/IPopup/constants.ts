/**
 * @internal
 */
export const MIN_DESKTOP_WIDTH = 640;

/**
 * Space between popup and anchor in px.
 *
 * @internal
 */
export const POPUP_SPACING = 20;
