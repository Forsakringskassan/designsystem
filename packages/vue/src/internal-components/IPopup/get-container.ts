import { config } from "../../config";

/**
 * Get container element from one of the following sources (in order):
 *
 * 1. An explicit element given by the container prop.
 * 2. A parent marked as a popup container.
 * 3. Default configured popup container element (e.g. `<body>`).
 */
export function getContainer(
    element: Element,
    prop: HTMLElement | null | undefined,
): HTMLElement {
    if (prop) {
        return prop;
    }
    const parent = element.closest<HTMLElement>(".popup__container");
    if (parent) {
        return parent;
    }
    return config.popupContainer;
}
