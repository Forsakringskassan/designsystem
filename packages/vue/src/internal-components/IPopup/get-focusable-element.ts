import { findTabbableElements } from "@fkui/logic";
import { getHTMLElementFromVueRef } from "../../utils";

export function getFocusableElement(
    rootElement: unknown,
    callback?: () => HTMLElement | null,
): HTMLElement | null {
    /* if user callback is provided use that only, if it returns null no element should be focused */
    if (callback) {
        return callback();
    }

    /* if no callback is provided we lookup the first tabbable element in the popup */
    const popupElement = getHTMLElementFromVueRef(rootElement);
    const elements = findTabbableElements(popupElement);
    return elements[0] ?? null; /* null coalece */
}
