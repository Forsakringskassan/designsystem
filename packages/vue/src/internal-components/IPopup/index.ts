export { default as IPopup } from "./IPopup.vue";
