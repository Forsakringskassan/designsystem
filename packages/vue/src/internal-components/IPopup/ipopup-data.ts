import { type StackHandle } from "@fkui/logic";
import { Placement } from "./IPopupUtils";

/**
 * @public
 */
export interface IPopupData {
    teleportDisabled: boolean;
    placement: Placement;
    focus: StackHandle | null;
    noCloseOnResize: boolean;
}
