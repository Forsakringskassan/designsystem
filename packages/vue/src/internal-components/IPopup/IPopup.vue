<script lang="ts">
import { defineComponent, type PropType } from "vue";
import { debounce, handleTab, pushFocus, popFocus } from "@fkui/logic";
import { config } from "../../config";
import { getHTMLElementFromVueRef } from "../../utils";
import { getElement, fitInsideArea, getScrollToPopup, Placement, CandidateOrder } from "./IPopupUtils";
import { getContainer } from "./get-container";
import { getFocusableElement } from "./get-focusable-element";
import { isTeleportDisabled } from "./is-teleport-disabled";
import { MIN_DESKTOP_WIDTH, POPUP_SPACING } from "./constants";
import { type IPopupData } from "./ipopup-data";

export default defineComponent({
    name: "IPopup",
    inheritAttrs: false,
    props: {
        /**
         * Toggle open/closed popup.
         */
        isOpen: {
            type: Boolean,
            required: true,
        },
        /**
         * DOM element to position popup at.
         */
        anchor: {
            type: HTMLElement as PropType<HTMLElement | null | undefined>,
            required: false,
            default: undefined,
        },
        /**
         * Type of inline behaviour.
         * - `"auto"` changes between overlay or inline depending on window size.
         * - `"always"` forces the popup to always be inline.
         * - `"never"` forces the popup to never be inline.
         */
        inline: {
            type: String as PropType<"always" | "never" | "auto">,
            required: false,
            validator(value: string) {
                // The value must match one of these strings
                return ["always", "never", "auto"].includes(value);
            },
            default: "auto",
        },
        /**
         * Which element to use as container.
         */
        container: {
            type: HTMLElement as PropType<HTMLElement | null | undefined>,
            required: false,
            default: undefined,
        },
        /**
         * Which element to use as viewport.
         */
        viewport: {
            type: HTMLElement as PropType<HTMLElement>,
            required: false,
            default(): HTMLElement {
                return document.documentElement;
            },
        },
        /**
         * Prevents tabbing outside of component.
         */
        keyboardTrap: {
            type: Boolean,
            required: false,
            default: true,
        },
        /**
         * Function that returns the element that will receive focus
         */
        focusElement: {
            type: Function as PropType<() => HTMLElement | null>,
            required: false,
            default: null,
        },
        /**
         * Set focus on first tabbable element (or element in the `focusElement` prop if provided) when opened.
         */
        setFocus: {
            type: Boolean,
            required: false,
            default: true,
        },
    },
    emits: [
        /**
         * Emitted when popup is visible and placement is fully calculated.
         */
        "open",
        /**
         * Emitted when clicked outside of popup.
         */
        "close",
    ],
    data(): IPopupData {
        return {
            teleportDisabled: false,
            placement: Placement.NotCalculated,
            focus: null,
        };
    },
    computed: {
        popupClasses(): string[] {
            const popupState = this.isInline ? ["popup--inline"] : ["popup--overlay"];
            return ["popup", ...popupState];
        },
        isInline(): boolean {
            let isInline = this.teleportDisabled || this.placement === Placement.Fallback;

            if (this.forceInline) {
                isInline = true;
            } else if (this.forceOverlay) {
                isInline = false;
            } else if (this.placement === Placement.NotCalculated && !this.isMobileSize()) {
                // Overlay is required to get accurate results from placement calculation.
                isInline = false;
            }

            return isInline;
        },
        forceInline(): boolean {
            return this.inline === "always";
        },
        forceOverlay(): boolean {
            return this.inline === "never";
        },
        teleportTarget() {
            return config.teleportTarget;
        },
    },
    watch: {
        isOpen: {
            immediate: true,
            handler(value: boolean): void {
                this.toggleIsOpen(value);
                if (value) {
                    const { placement, forceInline, forceOverlay } = this;
                    this.teleportDisabled = isTeleportDisabled({ window, placement, forceInline, forceOverlay });

                    // wait one tick so we dont get the click
                    // that launches the popup (await nextTick doesnt work here)
                    setTimeout(() => {
                        // verify that it's still open
                        if (this.isOpen) {
                            document.addEventListener("click", this.onDocumentClickHandler);
                            window.addEventListener("resize", this.onWindowResizeDebounced);
                        }
                    }, 0);
                } else {
                    document.removeEventListener("click", this.onDocumentClickHandler);
                    window.removeEventListener("resize", this.onWindowResizeDebounced);
                }
            },
        },
    },
    created() {
        this.onWindowResizeDebounced = debounce(this.onWindowResize, 100).bind(this);
    },
    unmounted() {
        // Clean up if unmounted but still opened
        document.removeEventListener("click", this.onDocumentClickHandler);
        window.removeEventListener("resize", this.onWindowResizeDebounced);
    },
    methods: {
        async toggleIsOpen(isOpen: boolean): Promise<void> {
            /* popup is closing */
            if (!isOpen) {
                this.placement = Placement.NotCalculated;

                /* restore focus */
                if (this.focus) {
                    popFocus(this.focus);
                    this.focus = null;
                }
                return;
            }

            /* popup is opening */
            // Wait for popup to show up
            await this.$nextTick();
            await this.calculatePlacement();
            this.applyFocus();
            this.$emit("open");
        },
        async calculatePlacement(): Promise<void> {
            const popup = getHTMLElementFromVueRef(this.$refs.popup);
            const wrapper = getHTMLElementFromVueRef(this.$refs.wrapper);
            const anchor = getElement(this.anchor);

            if (!anchor) {
                throw new Error("No anchor element found");
            }

            // Check candidates for overlay placement.
            const shouldCheckCandidates = this.forceOverlay || !(this.isMobileSize() || this.forceInline);
            if (shouldCheckCandidates) {
                const area = getContainer(popup, this.container);
                const viewport = this.viewport;
                const result = fitInsideArea({
                    area,
                    anchor,
                    target: wrapper,
                    viewport,
                    spacing: POPUP_SPACING,
                    candidateOrder: CandidateOrder.Default,
                });

                this.placement = result.placement;
                const useOverlay = this.forceOverlay || result.placement !== Placement.Fallback;
                if (useOverlay) {
                    wrapper.style.left = `${result.x}px`;
                    wrapper.style.top = `${result.y}px`;
                    return;
                }
            }

            // Force disable teleport since it won't be updated due to fallback placement.
            this.teleportDisabled = true;
            wrapper.style.removeProperty("left");
            wrapper.style.removeProperty("top");

            // Wait for virtual keyboard to be closed before calculating scroll.
            await new Promise((resolve) => setTimeout(resolve, 200));

            const scrollTarget = popup.closest(".scroll-target");
            const hasScrollTarget = scrollTarget !== null;
            const top = getScrollToPopup({
                popup: wrapper,
                windowInnerHeight: window.innerHeight,
                scrollTop: hasScrollTarget ? scrollTarget.scrollTop : window.scrollY,
                spacing: POPUP_SPACING,
            });
            const scrollOptions = { top, behavior: "smooth" } as const;

            if (hasScrollTarget) {
                scrollTarget.scrollTo(scrollOptions);
            } else {
                window.scrollTo(scrollOptions);
            }
        },
        applyFocus(): void {
            if (!this.setFocus) {
                return;
            }
            const wrapper = this.$refs.wrapper;
            if (!wrapper) {
                return;
            }
            const focusableElement = getFocusableElement(wrapper, this.focusElement);
            this.focus = pushFocus(focusableElement);
        },
        isMobileSize(): boolean {
            return window.innerWidth < MIN_DESKTOP_WIDTH;
        },
        onDocumentClickHandler(): void {
            this.$emit("close");
        },
        onWindowResizeDebounced(): void {
            // Overwritten in created so that the debounced `onWindowResize`
            // method can be removed by removeEventListener.
        },
        async onWindowResize(): Promise<void> {
            // Abort if popup was closed during debounce.
            if (!this.isOpen) {
                return;
            }
            // Don't need to recalculate on resize if forced inline.
            if (this.forceInline) {
                return;
            }
            // Don't need to recalculate if popup is still supposed to be inline.
            if (this.isInline && this.isMobileSize()) {
                return;
            }

            // Reset placement/teleport to get overlay state for accurate calculation.
            if (this.isInline) {
                this.placement = Placement.NotCalculated;
                this.teleportDisabled = false;
                await this.$nextTick();
            }

            await this.calculatePlacement();
            const { placement, forceInline, forceOverlay } = this;
            this.teleportDisabled = isTeleportDisabled({ window, placement, forceInline, forceOverlay });
        },
        onPopupClickHandler(event: Event): void {
            // prevent propagation so we don't catch this
            // event in onDocumentClickHandler and close ourself.
            event.stopPropagation();
        },
        onKeyEsc(): void {
            this.$emit("close");
        },
        onKeyTab(event: KeyboardEvent): void {
            if (this.keyboardTrap) {
                const wrapper = getHTMLElementFromVueRef(this.$refs.wrapper);
                handleTab(event, wrapper);
            }
        },
    },
});
</script>

<template>
    <teleport v-if="isOpen" :to="teleportTarget" :disabled="teleportDisabled">
        <div ref="popup" v-bind="$attrs" :class="popupClasses">
            <div
                ref="wrapper"
                role="presentation"
                class="popup__wrapper"
                @click="onPopupClickHandler"
                @keyup.esc.stop="onKeyEsc"
                @keydown.tab="onKeyTab"
            >
                <slot v-bind="{ toggleIsOpen, placement }"></slot>
            </div>
        </div>
    </teleport>
</template>
