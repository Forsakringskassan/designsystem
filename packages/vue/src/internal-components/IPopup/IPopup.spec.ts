import "html-validate/jest";
import "@fkui/test-utils/jest";
import { VueWrapper, mount } from "@vue/test-utils";
import { defineComponent } from "vue";
import flushPromises from "flush-promises";
import { createPlaceholderInDocument } from "@fkui/test-utils/vue";
import IPopup from "./IPopup.vue";

jest.useFakeTimers();

function createWrapper(
    template: string,
): VueWrapper<InstanceType<typeof TestComponent>> {
    const TestComponent = defineComponent({
        name: "TestComponent",
        components: {
            IPopup,
        },
        data() {
            return { isOpen: false, gotOpenEvent: false, gotCloseEvent: false };
        },
        template,
    });

    return mount(TestComponent, {
        attachTo: createPlaceholderInDocument(),
        global: {
            stubs: ["teleport"],
        },
    });
}

async function mountPopup(
    template: string = defaultTemplate,
): Promise<ReturnType<typeof createWrapper>> {
    const wrapper = createWrapper(template);
    await flushPromises();
    return wrapper;
}

async function openPopup(wrapper: VueWrapper): Promise<void> {
    await wrapper.get("#launch-popup").trigger("click");
    await jest.runAllTimers();
    await flushPromises();
}

const defaultTemplate = /* HTML */ `
    <div id="outside">
        <button id="launch-popup" ref="anchor" @click="isOpen=true"></button>
        <i-popup
            :isOpen="isOpen"
            :anchor="$refs.anchor"
            @open="gotOpenEvent = true"
            @close="isOpen = false; gotCloseEvent = true;"
        >
            <span> POPUP CONTENT </span>
        </i-popup>
    </div>
`;

beforeEach(() => {
    jest.restoreAllMocks();
});

describe("snapshots", () => {
    it("should match snapshot when open", async () => {
        jest.spyOn(window, "scrollTo").mockReturnValue();
        const wrapper = await mountPopup();
        await openPopup(wrapper);
        expect(wrapper.element).toMatchSnapshot();
    });

    it("should match snapshot when closed", async () => {
        const wrapper = await mountPopup();
        expect(wrapper.element).toMatchSnapshot();
    });
});

describe("events", () => {
    it('should emit "open" event after popup has opened', async () => {
        jest.spyOn(window, "scrollTo").mockReturnValue();

        const wrapper = await mountPopup();
        await openPopup(wrapper);

        expect(wrapper.vm.$data["gotOpenEvent"]).toBeTruthy();
    });

    it('should emit "close" event on escape key pressed', async () => {
        jest.spyOn(window, "scrollTo").mockReturnValue();

        const wrapper = await mountPopup();
        await openPopup(wrapper);

        const closeElement = wrapper.get(".popup__wrapper");
        await closeElement.trigger("keyup.esc");
        await wrapper.vm.$nextTick();
        await flushPromises();

        expect(wrapper.vm.$data["gotCloseEvent"]).toBeTruthy();
    });

    it('should emit "close" event when clicked outside an open popup', async () => {
        jest.spyOn(window, "scrollTo").mockReturnValue();

        const wrapper = await mountPopup();
        await openPopup(wrapper);

        await wrapper.get("#outside").trigger("click");
        await wrapper.vm.$nextTick();
        await flushPromises();

        expect(wrapper.vm.$data["gotCloseEvent"]).toBeTruthy();
    });

    it('should not emit "close" event when clicked outside a closed popup', async () => {
        const wrapper = await mountPopup();

        await wrapper.get("#outside").trigger("click");

        expect(wrapper.vm.$data["gotCloseEvent"]).toBeFalsy();
    });
});

describe("html-validate", () => {
    it("should require is-open attribute", () => {
        expect("<i-popup></i-popup>").not.toHTMLValidate({
            ruleId: "element-required-attributes",
            message: '<i-popup> is missing required "is-open" attribute',
        });
    });

    it("should require anchor attribute", () => {
        expect("<i-popup></i-popup>").not.toHTMLValidate({
            ruleId: "element-required-attributes",
            message: '<i-popup> is missing required "anchor" attribute',
        });
    });

    it("should only allow setting valid `inline` values", () => {
        expect(/* HTML */ `
            <i-popup anchor="anchorref" is-open inline="always"></i-popup>
        `).toHTMLValidate();
        expect(/* HTML */ `
            <i-popup anchor="anchorref" is-open inline="never"></i-popup>
        `).toHTMLValidate();
        expect(/* HTML */ `
            <i-popup anchor="anchorref" is-open inline="auto"></i-popup>
        `).toHTMLValidate();
        expect(/* HTML */ `
            <i-popup anchor="anchorref" is-open inline="foo"></i-popup>
        `).not.toHTMLValidate({
            ruleId: "attribute-allowed-values",
            message: 'Attribute "inline" has invalid value "foo"',
        });
    });

    it("should allow setting viewport value", () => {
        const markup = /* HTML */ `
            <i-popup
                anchor="anchorref"
                is-open
                viewport="viewportref"
            ></i-popup>
        `;
        expect(markup).toHTMLValidate();
    });

    it("should allow setting focus-element value", () => {
        const markup = /* HTML */ `
            <i-popup
                anchor="anchorref"
                is-open
                focus-element="focuselementref"
            ></i-popup>
        `;
        expect(markup).toHTMLValidate();
    });
});
