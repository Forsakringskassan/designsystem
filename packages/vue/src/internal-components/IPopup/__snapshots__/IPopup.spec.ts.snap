// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots should match snapshot when closed 1`] = `
<div
  id="outside"
>
  <button
    id="launch-popup"
  />
  <!--v-if-->
</div>
`;

exports[`snapshots should match snapshot when open 1`] = `
<div
  id="outside"
>
  <button
    id="launch-popup"
  />
  <teleport-stub
    disabled="false"
    to="[object HTMLBodyElement]"
  >
    <div
      class="popup popup--overlay"
    >
      <div
        class="popup__wrapper"
        role="presentation"
      >
        
        <span>
           POPUP CONTENT 
        </span>
        
      </div>
    </div>
  </teleport-stub>
</div>
`;
